src/NHibernate/Criterion/EntityProjectionType.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Data.Common;
+using NHibernate.Engine;
+using NHibernate.Event;
+using NHibernate.Persister.Entity;
+using NHibernate.Type;
+
+namespace NHibernate.Criterion
+{
+	/// <summary>
+	/// Specialized type for retrieving entity from Criteria/QueryOver API projection.
+	/// Intended to be used only inside <see cref="EntityProjection"/>
+	/// </summary>
+	[Serializable]
+	internal partial class EntityProjectionType : ManyToOneType, IType
+	{
+		private readonly EntityProjection _projection;
+
+		public EntityProjectionType(EntityProjection projection) : base(projection.RootEntity.FullName, projection.Lazy)
+		{
+			_projection = projection;
+		}
+
+		object IType.NullSafeGet(DbDataReader rs, string[] names, ISessionImplementor session, object owner)
+		{
+			//names parameter is ignored (taken from projection)
+			return NullSafeGet(rs, string.Empty, session, owner);
+		}
+
+		public override object NullSafeGet(DbDataReader rs, string name, ISessionImplementor session, object owner)
+		{
+			var identifier = _projection.Persister.IdentifierType.NullSafeGet(rs, _projection.IdentifierColumnAliases, session, null);
+
+			if (identifier == null)
+			{
+				return null;
+			}
+
+			return _projection.Lazy
+				? ResolveIdentifier(identifier, session)
+				: GetInitializedEntityFromProjection(rs, session, identifier);
+		}
+
+		private object GetInitializedEntityFromProjection(DbDataReader rs, ISessionImplementor session, object identifier)
+		{
+			var entity = CreateInitializedEntity(
+				rs,
+				session,
+				_projection.Persister,
+				identifier,
+				_projection.PropertyColumnAliases,
+				LockMode.None,
+				_projection.FetchLazyProperties,
+				_projection.IsReadOnly);
+
+			return entity;
+		}
+
+		private static object CreateInitializedEntity(DbDataReader rs, ISessionImplementor session, IQueryable persister, object identifier, string[][] propertyAliases, LockMode lockMode, bool fetchLazyProperties, bool readOnly)
```

Comment:
I guess all that has been deducted from the loader logic. I am uncomfortable with having it re-coded here. I wonder how eager loading works for other cases like HQL or Linq. We should try to re-use its implementations for those other cases.
(From how is implemented `ManyToOneType` and `EntityType`, I tend to think that from the query results something hydrates and put eagerly fetched entities into the session cache before the `ManyToOneType` is hydrated, allowing it to not query the db.)

Reply:
The thing is - I didn't find anything reusable. 

>I tend to think that from the query results something hydrates and put eagerly fetched entities into the session cache before the `ManyToOneType` is hydrated, allowing it to not query the db

I'm not sure I'm following. How this logic is related here - we are in the state when query is already executed. And we need to materialize our results to entities.
My current understanding `ManyToOneType` doesn't support loading of initialized entities. It only reads identifier from data reader. If entity needs to be fully initialized - separate query is executed (that's exact behavior of #302)

Reply:
> I'm not sure I'm following. How this logic is related here 
Ok - I got it now ) And got me thinking how my code would behave if some entities are already in session.

Reply:
> And got me thinking how my code would behave if some entities are already in session.

Good point, things would likely not be coherent in such case with current state of the PR. Normally NHibernate yields the entity already in the session, with the state it has already, disregarding the data from the query. So this is another reason to check how eager loading is implemented for other cases. (I have never checked that, I am not more knowledgeable than you on this subject.)